### PR TITLE
[MO] - upgrading test-container to be more flexible + multinode support

### PR DESCRIPTION
**Pull request by see-quick**
Signed-off-by: morsak <xorsak02@stud.fit.vutbr.cz>

### Type of change

- Enhancement / new feature
- Refactoring

### Description

This PR adding support for custom configuration of `strimzi-test-container`.  Kafka configs are injected by map which can be specified in the Container constructor. Moreover, I have removed dependency on EmbeddedKafkaCluster and replaced it with  `strimzi-test-container`.

### Checklist

- [x] Make sure all tests pass

### CODE REVIEW

**Review by scholzj — COMMENTED** [review]
Looks good. I think @tombentley and @sknot-rh  might know better about how needed is the multi-broker setup. 

Do you plan to do the same also for the User and Topic Operators?

**Comment by Frawless** [comment]
I think a multinode setup would be useful not just for us, but for the community as well.

File: test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java [hunk 1]
```diff
@@ -20,7 +20,9 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 public class StrimziKafkaContainer extends GenericContainer<StrimziKafkaContainer> {
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:27]: Can we add a sentence of two of Javadoc explaining what this is and how it's intended to be used?

File: test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java [hunk 2]
```diff
@@ -73,18 +76,24 @@
         LOGGER.info("Supported Strimzi version: {}", STRIMZI_VERSION);
     }
 
-    public StrimziKafkaContainer(final String version) {
+    public StrimziKafkaContainer(final String version, Map<String, String> additionalKafkaConfiguration) {
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:79]: Perhaps rename `version` to `imageVersion`, since that's what it seems to be (rather than a Kafka version string).

File: test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java [hunk 3]
```diff
@@ -133,13 +142,23 @@ protected void containerIsStarting(InspectContainerResponse containerInfo, boole
...
+        // default listener config
+        kafkaConfiguration.append(" --override listeners=" + kafkaListeners + "PLAINTEXT://0.0.0.0:" + KAFKA_PORT);
+        kafkaConfiguration.append(" --override advertised.listeners=" + advertisedListeners);
+        kafkaConfiguration.append(" --override zookeeper.connect=localhost:" + ZOOKEEPER_PORT);
+        kafkaConfiguration.append(" --override listener.security.protocol.map=" + kafkaListenerSecurityProtocol + "PLAINTEXT:PLAINTEXT");
+        kafkaConfiguration.append(" --override inter.broker.listener.name=BROKER1");
+
+        // additional kafka config
+        kafkaConfigurationMap.forEach((configName, configValue) -> {
+            kafkaConfiguration.append(" --override " + configName + "=" + configValue);
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:156]: Validate that none of the `configName` is one of the hard coded configs.

Also, this could be a `.map(...).forEach(kafkaConfiguration::append)`, I think.

File: test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java [hunk 4]
```diff
@@ -133,13 +142,23 @@ protected void containerIsStarting(InspectContainerResponse containerInfo, boole
...
         });
 
+        StringBuilder kafkaConfiguration = new StringBuilder();
+
+        // default listener config
+        kafkaConfiguration.append(" --override listeners=" + kafkaListeners + "PLAINTEXT://0.0.0.0:" + KAFKA_PORT);
+        kafkaConfiguration.append(" --override advertised.listeners=" + advertisedListeners);
+        kafkaConfiguration.append(" --override zookeeper.connect=localhost:" + ZOOKEEPER_PORT);
+        kafkaConfiguration.append(" --override listener.security.protocol.map=" + kafkaListenerSecurityProtocol + "PLAINTEXT:PLAINTEXT");
+        kafkaConfiguration.append(" --override inter.broker.listener.name=BROKER1");
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:152]: You could chain the appends together, and onto the `new StringBuilder`.

**Review by tombentley — COMMENTED** [review]
We should be able to update `TopicOperatorIT` to use this too, right? (Can be a separate PR)

**Comment by see-quick** [comment]
> I think a multinode setup would be useful not just for us, but for the community as well.

Your wishes come true 🧙‍♀️🦄

File: test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java (hunk 3 above)
Comment (see-quick) [reply test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:163]: I don't want to forbid users from these options. For instance, if the user wants to override`listeners` or `zookeeper.connect` config is ok or?

**Comment by see-quick** [comment]
I will do update for `TopicOperatorIT` in separate PR @tombentley

File: test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java (hunk 3 above)
Comment (tombentley) [reply test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:171]: Fair enough, but does `--override` support later options overriding previous ones?

Comment (see-quick) [reply test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:171]: Yes It does )

**Review by Frawless — APPROVED** [review]
I think it would be nice to write a few sentences about it (how to use it, etc), but maybe after the migration to the new repo.

**Comment by see-quick** [comment]
@Frawless in the brand new repository, which should be created as this PR will be merged I will do it. Moreover, I thought about some BlogPost to write about it...I'm wondering if it's a good idea.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectorIT.java [hunk 5]
```diff
@@ -67,24 +76,40 @@ public static void before() throws IOException {
...
         String connectClusterName = getClass().getSimpleName();
-        cluster.createTopics(connectClusterName + "-offsets", connectClusterName + "-config", connectClusterName + "-status");
+
+        CreateTopicsResult createTopicsResult = adminClient.createTopics(
+            Arrays.asList(
+                new NewTopic(connectClusterName + "-offsets", 3, (short) 3),
+                new NewTopic(connectClusterName + "-config", 3,  (short) 3),
+                new NewTopic(connectClusterName + "-status", 3,  (short) 3)
+            ));
+        createTopicsResult.all().get();
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectorIT.java:107]: Why is this needed? Shouldn't Connect create this automatically?

File: test-container/pom.xml [hunk 6]
```diff
@@ -23,6 +23,16 @@
...
+        <dependency>
+            <groupId>com.github.spotbugs</groupId>
+            <artifactId>spotbugs-annotations</artifactId>
+        </dependency>
+        <!-- for TestUtils.waifFor       -->
+        <dependency>
+            <groupId>io.strimzi</groupId>
+            <artifactId>test</artifactId>
+            <scope>compile</scope>
+        </dependency>
```
Comment (scholzj) [inline test-container/pom.xml:35]: This will not work well once we move it to the sepsrate repo.

Comment (see-quick) [reply test-container/pom.xml:35]: I know but should I insert also the implementation of `waitFor` in some utils?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectorIT.java (hunk 5 above)
Comment (see-quick) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectorIT.java:107]: I don't know I just re-write this code to be equivalent by instead of using Kafka cluster to create it I am using AdminClient. I can give it a try....and see what happens...

Comment (see-quick) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectorIT.java:107]: I see that @tombentley was first, which wrote that `cluster.createTopics(getClass().getSimpleName() + "-offsets", getClass().getSimpleName() + "-config", getClass().getSimpleName() + "-status");`... I tried to remove that code and test also passed so I don't really know :D 🤷‍♂️

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectorIT.java [hunk 7]
```diff
@@ -67,8 +70,11 @@ public static void before() throws IOException {
                         .setEnabled(true)
         ));
 
-        cluster = new EmbeddedKafkaCluster(3);
-        cluster.start();
+        kafkaCluster = new EmbeddedKafkaCluster(3);
+        kafkaCluster.start();
+
+        Properties properties = new Properties();
+        properties.setProperty(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaCluster.bootstrapServers());
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectorIT.java:77]: I guess this is not needed?

File: test-container/pom.xml [hunk 8]
```diff
@@ -23,6 +23,10 @@
             <groupId>org.testcontainers</groupId>
             <artifactId>testcontainers</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.github.spotbugs</groupId>
+            <artifactId>spotbugs-annotations</artifactId>
+        </dependency>
```
Comment (scholzj) [inline test-container/pom.xml:29]: Should this be set as provided so it is not pulled unnecessarily?

File: cluster-operator/pom.xml [hunk 9]
```diff
@@ -314,6 +279,51 @@
             <version>${mockserver.version}</version>
             <scope>test</scope>
         </dependency>
+        <!-- Used by EmbeddedKafkaCluster -->
```
Comment (scholzj) [inline cluster-operator/pom.xml:282]: Can we maybe move this to the original place to avoid unnecessary changes?

**Review by scholzj — APPROVED** [review]

**Comment by scholzj** [comment]
@tombentley Do you have something more to this?

File: test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    @Override
+    public void start() {
+        Stream<Startable> startables = this.brokers.stream().map(Startable.class::cast);
+        try {
+            Startables.deepStart(startables).get(60, TimeUnit.SECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            e.printStackTrace();
+        }
+
+        TestUtils.waitFor("Zookeeper readiness", Duration.ofSeconds(1).toMillis(), Duration.ofSeconds(30).toMillis(),
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java:112]: We're not waiting for "Zookeeper readiness" here. We're waiting for the broker znode.

File: test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+                    e.printStackTrace();
+                    return false;
+                }
+            });
+    }
+
+    @Override
+    public void stop() {
+        // stop all containers in parallel
+        Stream.concat(this.brokers.stream(), Stream.of(this.zookeeper))
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java:135]: On the command line if I kill ZK first it sometimes prevents a broker from shutting down quickly. So maybe it's better to do the ZK shudown first, and then shut down the brokers in parallel.

File: test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        LOGGER.info("This is mapped port {}", zookeeperExposedPort);
+
+        String command = "#!/bin/bash \n";
+
+        command += "bin/zookeeper-server-start.sh config/zookeeper.properties\n";
+
+        LOGGER.info("Copying command to 'STARTER_SCRIPT' script.");
+
+        copyFileToContainer(
+            Transferable.of(command.getBytes(StandardCharsets.UTF_8), 700),
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java:80]: What's the intent behind having this starter script, if it just calls `zookeeper-server-start.sh`?

File: test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (!super.equals(o)) return false;
+        StrimziZookeeperContainer that = (StrimziZookeeperContainer) o;
+        return zookeeperExposedPort == that.zookeeperExposedPort;
+    }
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), zookeeperExposedPort);
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java:95]: Including `super.hashCode` like this seems to result in a `hashCode()` that's incompatible with `equals()`.

File: test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        );
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (!super.equals(o)) return false;
+        StrimziZookeeperContainer that = (StrimziZookeeperContainer) o;
+        return zookeeperExposedPort == that.zookeeperExposedPort;
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java:91]: This isn't really sufficient for equality. I could have two instances, and the exposed port constraint means only one can run at once, but both instances can exist. Do we need these containers to override equals?

File: test-container/src/main/java/io/strimzi/utils/WaitException.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.utils;
+
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/utils/WaitException.java:6]: A line of Javadoc please, to explain what a WaitException is supposed to be conceptually.

File: test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java (hunk 11 above)
Comment (see-quick) [reply test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java:135]: good point yes

File: test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java (hunk 12 above)
Comment (see-quick) [reply test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java:80]: It will never call `just` zookeeper-server-start.sh. There are two paths:

- configured to use external zookeeper -> 1. kafka-server-start.sh
- configured to use embedded zookeeper -> 1. zookeeper-server-start.sh, 2. kafka-server-start.sh 

the whole idea of this STARTING script is because we need to change the advertised configuration of the Kafka container before it starts...

File: test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java (hunk 14 above)
Comment (see-quick) [reply test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java:91]: I think we don't need them and I could add spot bugs suppress annotation

**Comment by scholzj** [comment]
@see-quick I'm not completely sure if we should merge it without approval from @tombentley. Is it blocking you in some further work on the test container?

File: test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java [hunk 16]
```diff
@@ -94,13 +112,18 @@ protected void doStart() {
         super.doStart();
     }
 
+    public StrimziKafkaContainer withExternalZookeeper(String connectString) {
+        this.externalZookeeperConnect = connectString;
+        return self();
+    }
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziKafkaContainer.java:118]: Perhaps private constructors and public factory methods to separate the two flavours would be better that this mutable field needing to be set before starting the container?

File: test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import org.testcontainers.containers.Network;
+import org.testcontainers.images.builder.Transferable;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * StrimziZookeeperContainer is an instance of the Zookeeper encapsulated inside a docker container using the latest image
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/StrimziZookeeperContainer.java:21]: Similar rewording for the parametrised image version.

File: test-container/src/main/java/io/strimzi/utils/TestUtils.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+                waitException.printStackTrace();
+                throw waitException;
+            }
+            long sleepTime = Math.min(pollIntervalMs, timeLeft);
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("{} not ready, will try again in {} ms ({}ms till timeout)", description, sleepTime, timeLeft);
+            }
+            try {
+                Thread.sleep(sleepTime);
+            } catch (InterruptedException e) {
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/utils/TestUtils.java:73]: Probably better to throw `InterruptedException`.

File: test-container/src/main/java/io/strimzi/utils/TestUtils.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+            if (timeLeft <= 0) {
+                if (exceptionCount > 1) {
+                    LOGGER.error("Exception waiting for {}, {}", description, exceptionMessage);
+
+                    if (!stackTraceError.toString().isEmpty()) {
+                        // printing handled stacktrace
+                        LOGGER.error(stackTraceError.toString());
+                    }
+                }
+                onTimeout.run();
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/utils/TestUtils.java:62]: This whole thing with `stackTraceError` and the calls to `printStackTrace` is a bit horrible. It's not really your job here to decide what should be printed to stderr. Perhaps using suppressed exceptions would be better? Better if you only have a single suppressed exception though.

File: test-container/src/main/java/io/strimzi/utils/TestUtils.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+                WaitException waitException = new WaitException("Timeout after " + timeoutMs + " ms waiting for " + description);
+                waitException.printStackTrace();
+                throw waitException;
+            }
+            long sleepTime = Math.min(pollIntervalMs, timeLeft);
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("{} not ready, will try again in {} ms ({}ms till timeout)", description, sleepTime, timeLeft);
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/utils/TestUtils.java:69]: Perhaps "satisfied" is better than "ready"?

File: test-container/src/test/java/io/strimzi/StrimziKafkaClusterTest.java [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+
+        Map<String, String> kafkaClusterConfiguration = new HashMap<>();
+        kafkaClusterConfiguration.put("zookeeper.connect", "zookeeper:2181");
+
+        systemUnderTest = new StrimziKafkaCluster(StrimziKafkaContainer.getStrimziVersion() + "-kafka-" + StrimziKafkaContainer.getLatestKafkaVersion(),
+            numberOfBrokers,
+            numberOfReplicas,
+            kafkaClusterConfiguration);
+        systemUnderTest.start();
+
```
Comment (tombentley) [inline test-container/src/test/java/io/strimzi/StrimziKafkaClusterTest.java:62]: This is identical for both tests -> `@BeforeEach`

File: test-container/src/test/java/io/strimzi/StrimziKafkaClusterTest.java [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+        );
+
+        final String brokers = result.getStdout();
+        // verify that all kafka brokers are bind to zookeeper
+        assertThat(brokers, notNullValue());
+        assertThat(brokers.split(",").length, is(numberOfBrokers));
+
+        LOGGER.info("Brokers are {}", systemUnderTest.getBootstrapServers());
+
+        systemUnderTest.stop();
```
Comment (tombentley) [inline test-container/src/test/java/io/strimzi/StrimziKafkaClusterTest.java:76]: `@AfterEach`

File: test-container/src/test/java/io/strimzi/StrimziKafkaClusterTest.java [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+                    return true;
+                });
+
+            consumer.unsubscribe();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } catch (ExecutionException e) {
+            e.printStackTrace();
+        } catch (TimeoutException e) {
+            e.printStackTrace();
```
Comment (tombentley) [inline test-container/src/test/java/io/strimzi/StrimziKafkaClusterTest.java:155]: Calling `printStackTrace` is almost never the right thing to do, it's usually a sign that exception handling isn't really being done properly. In this case, because it's a test, just declare `throws Exception`.

File: test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+
+        // multi-node set up
+        this.brokers = IntStream
+            .range(0, this.brokersNum)
+            .mapToObj(brokerId -> {
+                LOGGER.info("Starting broker with id {}", brokerId);
+                // adding broker id for each kafka container
+                additionalKafkaConfiguration.put("broker.id", String.valueOf(brokerId));
+
+                return new StrimziKafkaContainer(imageVersion, additionalKafkaConfiguration)
```
[the remark this answers is not in the document]
Comment (see-quick) [reply test-container/src/main/java/io/strimzi/StrimziKafkaCluster.java]: Done

File: test-container/src/main/java/io/strimzi/cluster/StrimziKafkaCluster.java [hunk 25]
```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.cluster;
```
[the remark this answers is not in the document]
Comment (see-quick) [reply test-container/src/main/java/io/strimzi/cluster/StrimziKafkaCluster.java]: > Sorry, I should have been more specific before, let's use something like io.strimzi.test.container.

Done

File: test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+public class StrimziKafkaCluster implements Startable {
+
+    private static final Logger LOGGER = LogManager.getLogger(StrimziKafkaCluster.class);
+
+    private final int brokersNum;
+    private final Network network;
+    private final StrimziZookeeperContainer zookeeper;
+    private final Collection<StrimziKafkaContainer> brokers;
+
+    @SuppressFBWarnings("BC_UNCONFIRMED_CAST_OF_RETURN_VALUE")
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java:44]: Do we actually need this?

File: test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+                return StrimziKafkaContainer.createWithAdditionalConfiguration(brokerId, imageVersion, additionalKafkaConfiguration)
+                    .withNetwork(this.network)
+                    .withNetworkAliases("broker-" + brokerId)
+                    .dependsOn(this.zookeeper)
+                    .withExternalZookeeper("zookeeper:" + StrimziZookeeperContainer.ZOOKEEPER_PORT);
+            })
+            .collect(Collectors.toList());
+    }
+
+    public StrimziKafkaCluster(Map<String, String> additionalKafkaConfiguration) {
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java:84]: Is this actually used?

File: test-container/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 28]
```diff
@@ -20,20 +20,33 @@
...
-    private static final int KAFKA_PORT = 9092;
-    private static final int ZOOKEEPER_PORT = 2181;
     private static final String LATEST_KAFKA_VERSION;
     private static final List<String> SUPPORTED_KAFKA_VERSIONS = new ArrayList<>(5);
     private static final String STRIMZI_VERSION;
 
+    private Map<String, String> kafkaConfigurationMap;
+    private String externalZookeeperConnect;
     private int kafkaExposedPort;
+    private int brokerId;
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:49]: This field is never used.

File: test-container/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java [hunk 29]
```diff
@@ -94,13 +118,18 @@ protected void doStart() {
         super.doStart();
     }
 
+    public StrimziKafkaContainer withExternalZookeeper(String connectString) {
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/test/container/StrimziKafkaContainer.java:121]: When I made the comment here before about the factory methods I was expecting the external ZK to be one of them. Now I see that you've got three overloads for those methods. Another 3 seems a bit silly, but maybe the external ZK should just be extra parameter for a 4th factory?

File: test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+                        "sh", "-c",
+                        "bin/zookeeper-shell.sh zookeeper:" + StrimziKafkaContainer.ZOOKEEPER_PORT + " ls /brokers/ids | tail -n 1"
+                    );
+                    String brokers = result.getStdout();
+
+                    LOGGER.info("Stdout from zookeeper container....{}", result.getStdout());
+
+                    return brokers != null && brokers.split(",").length == this.brokersNum;
+                } catch (IOException | InterruptedException e) {
+                    e.printStackTrace();
```
Comment (tombentley) [inline test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java:124]: Remind me again why it's OK to just ignore the `InterruptedException`? I know it's an annoying Exception to have to deal with, and I know you're not expecting to need to support cancellation here, but you never know how code will evolve and be reused. It would be better to do `Thread.currentThread().interrupt();`, that way `waitFor`'s call to `sleep()` would throw `InterruptedException` and we can declare `waitFor() throws InterruptedException`, and then the whole thing supports cancellation.

File: test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java (hunk 26 above)
Comment (see-quick) [reply test-container/src/main/java/io/strimzi/test/container/StrimziKafkaCluster.java:44]: No, we don't... Thanks

**Review by tombentley — APPROVED** [review]
